### PR TITLE
Avoid refreshing the whole test details page on job state transitions

### PR DESCRIPTION
**Pull request by Martchus**
See https://progress.opensuse.org/issues/32611

### CODE REVIEW

File: assets/javascripts/running.js [hunk 1]
```diff
@@ -8,17 +8,38 @@ var liveViewElements = [];
 // (initialized in initLivelogAndTerminal())
 var logElements;
 
+// Reload broken thumbnails (that didn't exist yet when being requested) every 7th time
```
Comment (kalikiana) [inline assets/javascripts/running.js:11]: Why every seventh time? Also, the previous code said every 7 seconds. I'm not sure either makes sense to me.

Comment (Martchus) [reply assets/javascripts/running.js:11]: I'm just outsourcing this code into a separate function. I actually prefer to focus on one aspect of the page at a time and not try to improve everything at once.

I've adjusted the comment because it was incorrect. This function ensures by no means that the broken thumbnails are reloaded every 7 seconds. They are just reloaded every 7th time the code reaches that point. This is very arbitrary.

There are also further problems in the code below. I haven't fixed them for the same reason and just aimed for not making things worse. I'll add these to @okurz "collection ticket".

Comment (Martchus) [reply assets/javascripts/running.js:11]: And the additional call to that function my change introduces does not make use of that weird "7th time" condition. So I don't think I'm making things worse here.

Comment (kalikiana) [reply assets/javascripts/running.js:11]: Sounds fair.

**Comment by okurz** [comment]
+1

**Comment by Martchus** [comment]
Seems like there are still TODOs:

[ ] Fix the fullstack test
[ ] Fix `test_result.js 195:367 Uncaught TypeError: Cannot set property 'selectedIndex' of null`
[ ] Update favicon

File: lib/OpenQA/WebAPI/Controller/Test.pm [hunk 2]
```diff
@@ -223,8 +223,8 @@ sub _stash_job_and_module_list {
 sub details {
     my ($self) = @_;
 
-    $self->_stash_job_and_module_list or return $self->reply->not_found;
-    $self->render('test/details');
+    my $job = $self->_stash_job_and_module_list or return $self->reply->not_found;
+    $self->render($job->should_show_autoinst_log ? 'test/autoinst_log_within_details' : 'test/details');
```
Comment (kalikiana) [inline lib/OpenQA/WebAPI/Controller/Test.pm:227]: That's neat actually. I saw the file at first wondering why you were splitting that out.

File: assets/javascripts/running.js [hunk 3]
```diff
@@ -318,14 +339,60 @@ function initLivestream() {
...
+        method: 'GET',
+        success: function(response) {
+            infoPanel.innerHTML = response;
+            const infoBoxContent = document.getElementById('info-box-content');
+            if (!infoBoxContent) {
+                return;
+            }
+            // update favicon and class of info panel
+            ['16', 'svg'].forEach(function(iconType) {
+                document.getElementById('favicon-' + iconType).href = infoBoxContent.dataset['faviconUrl-' + iconType];
```
Comment (kalikiana) [inline assets/javascripts/running.js:360]: Nice.

File: t/full-stack.t [hunk 4]
```diff
@@ -390,6 +389,14 @@ subtest 'Cache tests' => sub {
...
     ok wait_for_result_panel($driver, qr/Result: incomplete/), 'test 8 is incomplete';
 
+    subtest 'log shown within details tab (without page reload)' => sub {
+        $driver->find_element_by_link_text('Details')->click();
+        wait_for_ajax(msg => 'autoinst log embedded within "Details" tab');
+        ok(my $log = $driver->find_element('.embedded-logfile'), 'element for embedded logfile present');
+        wait_for_ajax(msg => 'log contents loaded');
+        like($log->get_text(), qr/Result: setup failure/, 'log contents present');
+    };
+
```
Comment (okurz) [inline t/full-stack.t:399]: Don't we have that covered in another test already? Do we really need that in the full stack test?

File: t/full-stack.t [hunk 5]
```diff
@@ -399,6 +406,7 @@ subtest 'Cache tests' => sub {
 
     like($log_content, qr/Failed to download.*non-existent.qcow2/, 'Test 8 failure message found in log');
     like($log_content, qr/Result: setup failure/,                  'Test 8 state correct: setup failure');
+
```
Comment (okurz) [inline t/full-stack.t:409]: that I see as an unrelated change

**Comment by okurz** [comment]
… or so I thought. Error from the webui process:

```
Use of uninitialized value in join or string at /home/okurz/local/os-autoinst/openQA/script/../lib/OpenQA/WebAPI/Controller/Test.pm line 361.
```

could very well be not specific to your changes though.

File: t/full-stack.t (hunk 4 above)
Comment (Martchus) [reply t/full-stack.t:399]: > Don't we have that covered in another test already?

Unfortunately not. We only have it covered when the log file is already available at the time the page is loaded. To avoid the reload the file is loaded dynamically and that needs testing.

> Do we really need that in the full stack test?

This way I don't need to add an additional complex test setup (either complex because of all involved components or because of massive mocking). And adding the additional check to an already existing test seems beneficial for keeping the increased test execution time to a minimum. The test also already has checks for the autoinst log anyways so it doesn't seem to odd as well.

**Comment by Martchus** [comment]
In which situation did you see the warning? It is in the controller code for the "Next & Previous" tab which I haven't changed so I don't think it is specific to the PR.

File: t/full-stack.t (hunk 5 above)
Comment (Martchus) [reply t/full-stack.t:409]: When removing this blank line, can I remove the blank line between the `like` calls? It makes most sense to either split this block or not split it at all. And it is not that unrelated considering I'm improving the fullstack test here. I can make it a separate commit, though.

**Comment by Martchus** [comment]
Looking further at the code I don't think the warning has to do anything with my changes and it will always be printed if a test misses the `MACHINE` setting.

File: t/full-stack.t (hunk 4 above)
Comment (okurz) [reply t/full-stack.t:396]: yes, sure, makes sense. Just wanted to crosscheck.

**Comment by okurz** [comment]
also https://build.opensuse.org/public/build/devel:openQA:TestGithub:PR-2988/openSUSE_Tumbleweed/x86_64/openQA/_log failed in

```
[ 1456s]     not ok 6 - POST /api/v1/jobs/99963/artefact
[ 1456s]     not ok 7 - 200 OK
[ 1456s]     1..7
[ 1456s] not ok 39 - upload asset: successful chunk upload
```

**Comment by Martchus** [comment]
I hope we can merge the PR this time. At least the test failure in the package build seems completely unrelated considering what I've changed. It is also preceded by `# Premature connection close` which looks like a race condition:

```
[ 1456s]     # Premature connection close
[ 1456s] 
[ 1456s]     #   Failed test 'POST /api/v1/jobs/99963/artefact'
[ 1456s]     #   at ./t/api/04-jobs.t line 339.
[ 1456s] 
[ 1456s]     #   Failed test '200 OK'
[ 1456s]     #   at ./t/api/04-jobs.t line 339.
[ 1456s]     #          got: undef
[ 1456s]     #     expected: '200'
[ 1456s]     # Looks like you failed 2 tests of 7.
[ 1456s] 
[ 1456s] #   Failed test 'upload asset: successful chunk upload'
```
